### ISSUE TITLE
Golang generated code is not working as expected

### ISSUE THREAD

**Issue by kpym** [firasdib/Regex101#1353]
## Code Generator Language

Go

### Code Snippet

I want to replace `A(.)C` in `ABC` with `C${1}A` (the result should be `CBA`). The current code is 

```golang

package main

import (
    "regexp"
    "fmt"
)

func main() {
    var re = regexp.MustCompile(`^A(.)C$`)
    var str = []byte(`ABC`)
    var substitution = []byte("C${1}A")
    var i = 0
    var count = 1
    
    str = re.ReplaceAllFunc(str, func(s []byte) []byte {
        if count < 0 {
            return substitution
        } else if i < count {
            i += 1
            
            return substitution
        }
        
        return s
    })
    
    fmt.Println(string(str))
}

```

which is not working.

The desired code is 

```golang
package main

import (
    "regexp"
    "fmt"
)

func main() {
    var re = regexp.MustCompile(`^A(.)C$`)
    var str = []byte(`ABC`)
    var substitution = []byte("C${1}A")
    
    str = re.ReplaceAll(str, substitution)
    
    fmt.Println(string(str))
}
```

**Comment by working-name** [comment]
Hi! Can you please elaborate on "not working"? Do you get an error/message? (I've never done anything in go)

**Comment by kpym** [comment]
You can check the result (`C${1}A`) in the [online playground](https://play.golang.org/p/0DhAqjF5VtR).

The code is doing "string" replacement, not "regex" one.

You can also check the code that I propos in the same [online playground](https://play.golang.org/p/9Py-UrrQiOJ).

**Comment by firasdib** [comment]
Hmm, it looks like you've compiled a non-global regex, is that correct? In that scenario, only the first occurrence should be replaced. The first code examples (tries) to do that, while the second does not (unless I'm misinterpreting something).

Perhaps you can help correct the whole code generator for golang, if there are cases where it is incorrect? You can view the four different scenarios here:

Non-global substitution: https://regex101.com/r/rRAG2X/1/codegen?language=golang
Global substitution: https://regex101.com/r/rRAG2X/2/codegen?language=golang
Non-global match: https://regex101.com/r/rRAG2X/3/codegen?language=golang
Global match: https://regex101.com/r/rRAG2X/4/codegen?language=golang

Thanks in advance!

**Comment by kpym** [comment]
You are right, the only problem is in the case of non-global substitution. Here is my proposal.

```go
package main

import (
	"fmt"
	"regexp"
)

func main() {
	var re = regexp.MustCompile(`A(.)C`)
	var str = "ABCABC"
	var substitution = `C${1}A`
	var count = 1 // nagative counter is equivalent to global case (replace all)

	str = re.ReplaceAllStringFunc(str, func(s string) string {
		if count == 0 {
			return s
		}

		count -= 1
		return re.ReplaceAllString(s, substitution)
	})

	fmt.Println(string(str))
}
```

What I have done : 
- I use `ReplaceAllStringFunc` in place of `ReplaceAllFunc` so we can use strings and not `[]byte` like in the other three cases.
- I return `re.ReplaceAllString(s,substitution)` in place of `substitution` because if `substitution` contains group identifiers (like `$1`) they must be expanded.
- I use string literals (\`...\`) only for the regular expressions but use a regular string ("...") for the `str` variable. I think that this is something that should de corrected in all four cases (not only in the non-global replacement case).
- I have simplified the code by removing the `i` variable.

You can check and play with this code in the [golang playground](https://play.golang.org/p/a1gZLftDRnz).

**Comment by kpym** [comment]
I have made a [proposal](https://github.com/golang/go/issues/40198) to manage this case easily in Go, but it was not accepted (for backward compatibility reasons). So putting the right code generator may help Go developers to write non-global replace properly.

**Comment by firasdib** [comment]
I will try to get around to this soon.

**Comment by kpym** [comment]
@firasdib thanks. Here is an alternative code that introduce the missing `ReplaceString` function with limiting parameter `n`. 

```go
package main

import (
	"fmt"
	"regexp"
)

func ReplaceString(src, search, repl string, n int) string {

	var re = regexp.MustCompile(search)

	src = re.ReplaceAllStringFunc(src, func(s string) string {
		if n == 0 {
			return s
		}

		n -= 1
		return re.ReplaceAllString(s, repl)
	})

	return src
}

func main() {
	var re = `A(.)C`
	var str = "ABCABC"
	var substitution = `C${1}A`

	fmt.Println(ReplaceString(str, re, substitution, 1))
}
```

You can use this one or the code from my [previous comment](https://github.com/firasdib/Regex101/issues/1353#issuecomment-658008523).

**Comment by Ouims** [comment]
Did you ever fix this @firasdib ?

**Comment by firasdib** [comment]
I believe this is fixed. If the current code generator still produces incorrect output, please reopen.
